comit/src/order.rs

```diff
@@ -89,27 +84,93 @@ impl BtcDaiOrder {
             created_at: OffsetDateTime::now_utc(),
         }
     }
+}
+
+/// A newtype representing a quantity in a certain base currency B.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Quantity<B> {
+    value: B,
+}
+
+impl Quantity<Bitcoin> {
+    pub fn new(value: Bitcoin) -> Self {
+        Self { value }
+    }
+
+    pub fn sats(&self) -> u64 {
+        self.value.as_sat()
+    }
+
+    /// The [`Bitcoin`] type encapsulates sats and btc well, hence we can just
+    /// provide access to the inner value here.
+    pub fn to_inner(&self) -> Bitcoin {
+        self.value
+    }
+}
+
+#[cfg(test)]
+pub fn btc(btc: f64) -> Quantity<Bitcoin> {
+    Quantity::new(Bitcoin::from_sat(
+        bitcoin::Amount::from_btc(btc).unwrap().as_sat(),
+    ))
+}
+
+/// A newtype representing the price of one unit of the base currency B in the
+/// quote currency Q.
+///
+/// The core idea around of this type is to enforce the unit of the rate when
+/// doing calculations. We achieve that by adding "loud" constructors and
+/// accessors for combinations of base and quote currency.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Price<B, Q> {
```

Comment:
Perhaps re-arranging the order of the type parameters would make this type read easier. Since it is sats/wei it seems more logical to have `<Q, B>`.

Reply:
So the price is actually wei/sat but I thought that base is more important and hence put that one first. When talking about forex, base + quote seems to be the more natural ordering, no?

Reply:
Funny you should mention it because a bunch of places in code use the field order 
```
quantity: Foo
price: Bar
```
and it has been bothering me, did that order get chosen because it matches base/quote. I would call the 'natural' order of those two fields price/quantity because A) price is more important B) p comes before q alphabetically C) That's how I say them

Obviously point C does not carry that much weight.

Back to the point though, in code `Price<asset::Bitcoin, Erc20Quantity>` reads uncomfortably for me where as `Price<Erc20Quantity, asset::Bitcoin>` reads better because Erc20Quantity comes first and that's the unit of price we use. Its not a super important though.

Reply:
I am not too fussed about the ordering to be honest so I will not object if someone chooses to reorder them but I am too lazy now to be honest :sweat_smile: